File: Mar 29 287. Find the Duplicate Number (Medium)/Solution.go
/* Counter Set O(N) | O(N) */
func findDuplicate(nums []int) int {
    counter := make(map[int]bool)
    for _, num := range(nums){
        if _, found := counter[num]; found{
            return num
        }
        counter[num] = true
    }
    return -1
}

// /* Negative Marking O(N) | O(1) */
func findDuplicate(nums []int) int {
    for _, num := range(nums){
        // abs for num
        if num < 0 {
            num *= - 1 
        }
        // Find targetIdx
        targetIdx := num - 1
        // If already negative means duplicate
        if nums[targetIdx] < 0{
            return targetIdx + 1
        }
        // Make neagative
        nums[targetIdx] *= -1
    }
    return -1
}

/* Binary Search (NlogN) | O(1) */
func findDuplicate(nums []int) int {
    left, right := 0, len(nums) - 1
    duplicate := 0
    for left <= right{
        mid := (left + right) / 2
        count := 0
        // Count numbers <= mid
        for _,num := range(nums){
            if num <= mid{
                count++
            }
        }
        // Search left
        if count > mid{
            duplicate = mid
            right = mid - 1
        } else { // Search right
            left = mid +1
        }      
    }
    return duplicate
}

File: Mar 29 287. Find the Duplicate Number (Medium)/Solution.py
# Counter Set O(N) | O(N)
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        counter = set()
        for num in nums:
            if num in counter:
                return num
            counter.add(num)
            
# Negative Marking O(N) | O(1)
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        for num in nums:
            # Find targetIdx
            targetIdx = abs(num) - 1
            # If already negative means duplicate
            if nums[targetIdx] < 0:
                return targetIdx + 1
            else:
                nums[targetIdx] *= -1
# Binary Search (NlogN) | O(1)     
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1
        
        while left <= right:
            mid = (left + right) // 2
            # Count numbers <= mid
            count = sum(num <= mid for num in nums)
            if count > mid:
                duplicate = mid
                right = mid - 1
            else:
                left = mid + 1
        return duplicate
